src/server/info_collector.h

```diff
@@ -177,15 +177,16 @@ class info_collector
     uint32_t _storage_size_fetch_interval_seconds;
     uint32_t _storage_size_retry_wait_seconds;
     uint32_t _storage_size_retry_max_count;
-    std::string _hotspot_detect_algorithm;
     ::dsn::task_ptr _storage_size_stat_timer_task;
     ::dsn::utils::ex_lock_nr _capacity_unit_update_info_lock;
     // mapping 'node address' --> 'last updated timestamp'
     std::map<std::string, string> _capacity_unit_update_info;
-    std::map<std::string, hotspot_calculator *> _hotspot_calculator_store;
-
-    hotspot_calculator *get_hotspot_calculator(const std::string &app_name,
-                                               const int partition_num);
+    // _hotspot_calculator_store is to save hotspot_partition_calculator for each table, a
+    // hotspot_partition_calculator saves historical hotspot data and alert perf_counters of
+    // corresponding table
+    std::map<std::string, std::shared_ptr<hotspot_partition_calculator>> _hotspot_calculator_store;
```

Comment:
Why didn't you use `std::unique_ptr` instead of `std::shared_ptr`?

Reply:
`hotspot_partition_calculator` both used in _hotspot_calculator_store and info_collector

Reply:
Yes, I see it. But only `_hotspot_calculator_store` holds it, which means has the right to determine its life cycle.
you can return a `hotspot_partition_calculator &` in the interface of  `_hotspot_calculator_store` for  the using of `info_collector`

Reply:
`share_ptr` is easier to achieve I think, it's not sensitive to performance here